pkg/utils/retry.go

```diff
@@ -39,7 +43,22 @@ func NewRetryRunner(maxTimes int, delay time.Duration, expBackoff bool) *retryRu
 	}
 }
 
+// NewProbabilisticRetryRunner returns a new runner that sets delay to time.Duration(rand.Int63n(f.interval)+1) * time.Millisecond
+func NewProbabilisticRetryRunner(maxTimes int, interval int64, expBackoff bool) *retryRunner {
+	return &retryRunner{
+		delay:         0,
+		expBackoff:    expBackoff,
+		maxTimes:      maxTimes,
+		probabilistic: true,
+		interval:      interval,
+		logger:        flogging.MustGetLogger("retry-runner"),
+	}
+}
+
 func (f *retryRunner) nextDelay() time.Duration {
+	if f.probabilistic && f.delay == 0 {
```

Comment:
We could have this in the constructor?

Reply:
I wanted to avoid to put this in the contractor because if there is a success at the first trial, then we don't need to sample randomness that might be expensive.

Reply:
This check is performed every time we call `nextDelay`, so several times per `Run`. But the constructor is created once per service and is then reuse for multiple invocations of `Run`

Reply:
but this check is cheaper than sampling randomness, no?

Reply:
I changed strategy, now it should be better :)